scalagen/src/main/scala/org/scalameta/scalagen/GeneratorApplication.scala

```diff
@@ -0,0 +1,74 @@
+package org.scalameta.scalagen
+
+import scala.meta._
+import scala.meta.contrib._
+import scala.meta.gen._
+
+/**
+  * All transformations, manipulations etc. should completely noop, by default.
+  *
+  * They should return the exact identical object thats inputed.
+  *
+  * Structural equality is slow, thus we want to remain as efficient as possible here.
```

Comment:
it's possible to implement much faster structural equality than what we currently have, it will still be nut slow compared to referential equality. `productPrefix` and `productIterator` and `scala.runtime.Statics` should do the job.

Reply:
another undocumented point is that `Tree.origin` (private[meta]) contains the position and is preserved during tree transforms, this can also be used to test equality.